src/base/bittorrent/infohash.h

```diff
@@ -30,39 +30,101 @@
+
         InfoHash() = default;
-        InfoHash(const lt::sha1_hash &nativeHash);
-        InfoHash(const QString &hashString);
         InfoHash(const InfoHash &other) = default;
 
+        InfoHash(const UnderlyingHashType &nativeHash)
+            : m_valid {true}
+            , m_nativeHash {nativeHash}
+        {
+            const QByteArray raw = QByteArray::fromRawData(nativeHash.data(), length());
+            m_hashString = QString::fromLatin1(raw.toHex());
+        }
+
         static constexpr int length()
         {
-            return lt::sha1_hash::size();
+            return UnderlyingHashType::size();
         }
 
-        bool isValid() const;
+        bool isValid() const
+        {
+            return m_valid;
+        }
 
-        operator lt::sha1_hash() const;
-        operator QString() const;
+        operator UnderlyingHashType() const
+        {
+            return m_nativeHash;
+        }
+
+        static InfoHash fromString(const QString &hashString)
+        {
+            if (hashString.size() != (length() * 2))
+                return {};
+
+            const QByteArray raw = QByteArray::fromHex(hashString.toLatin1());
+            if (raw.size() != length())  // QByteArray::fromHex() will skip over invalid characters
+                return {};
+
+            InfoHash result;
+            result.m_valid = true;
+            result.m_hashString = hashString;
+            result.m_nativeHash.assign(raw.constData());
+
+            return result;
+        }
+
+        QString toString() const
+        {
+            return m_hashString;
+        }
 
     private:
         bool m_valid = false;
-        lt::sha1_hash m_nativeHash;
+        UnderlyingHashType m_nativeHash;
         QString m_hashString;
     };
 
-    bool operator==(const InfoHash &left, const InfoHash &right);
-    bool operator!=(const InfoHash &left, const InfoHash &right);
-    bool operator<(const InfoHash &left, const InfoHash &right);
-    uint qHash(const InfoHash &key, uint seed);
+    template <std::ptrdiff_t N>
+    bool operator==(const InfoHash<N> &left, const InfoHash<N> &right)
+    {
+        return (static_cast<typename InfoHash<N>::UnderlyingHashType>(left)
+                == static_cast<typename InfoHash<N>::UnderlyingHashType>(right));
+    }
+
+    template <std::ptrdiff_t N>
+    bool operator!=(const InfoHash<N> &left, const InfoHash<N> &right)
+    {
+        return !(left == right);
+    }
+
+    template <std::ptrdiff_t N>
+    bool operator<(const InfoHash<N> &left, const InfoHash<N> &right)
+    {
+        return static_cast<typename InfoHash<N>::UnderlyingHashType>(left) < static_cast<typename InfoHash<N>::UnderlyingHashType>(right);
+    }
+
+    template <std::ptrdiff_t N>
+    uint qHash(const InfoHash<N> &key, const uint seed)
+    {
+        return ::qHash((std::hash<typename InfoHash<N>::UnderlyingHashType> {})(key), seed);
+    }
+
+    using SHA1Hash = InfoHash<160>;
+    using SHA256Hash = InfoHash<256>;
```

Comment:
You misunderstood the relations of sha* hashes and "info hash". 
From https://www.bittorrent.org/beps/bep_0052.html:
>infohash
The infohash is calculated by applying a hash function to the bencoded form of the info dictionary, which is a substring of the metainfo file. For meta version 2 SHA2-256 is used.

So info hash has versions and uses the respective SHA hashes (SHA1, SHA2-256), not the other way around.

Reply:
You probably mean the name of template.
What name do you propose? HashValue? Or Digest32 (by analogy with libtorrent)?

Reply:
Or maybe SHAHash?

Reply:
>Or Digest32 (by analogy with libtorrent)?

This would be acceptable.